### PR TITLE
Windows: Refactor container

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com
@swernli

This PR is part of the proposal described in issue 10662 to port the docker daemon to Windows. This PR is one of the more complex PRs unfortunately, but many future PRs will depend on this PR, so hopefully this one can be merged relatively quickly :)  This PR is the first step at forking container.go into Linux and Windows versions. In doing that, it also required forking volumes.go to get compilation and integration tests to succeed, so I have also pulled in some minor changes which were slated for future PRs too in volumes. Note that container_windows.go still contains TODOs which will be addressed in future PRs which depend on this PR (didn't want to make a huge PR affecting dozens of files).

(Rebased 4/30 after engine removed)

### CODE REVIEW

**Comment by crosbymichael** [comment]
LGTM

File: daemon/daemon.go [hunk 1]
```diff
@@ -154,11 +154,11 @@ func (daemon *Daemon) containerRoot(id string) string {
...
 func (daemon *Daemon) load(id string) (*Container, error) {
-	container := &Container{
-		root:         daemon.containerRoot(id),
-		State:        NewState(),
-		execCommands: newExecStore(),
-	}
+	container := &Container{}
+	container.root = daemon.containerRoot(id)
+	container.State = NewState()
+	container.execCommands = newExecStore()
```
Comment (tiborvass) [inline daemon/daemon.go]: @jhowardmsft why this change? I actually prefer the way it was :S

File: daemon/daemon.go [hunk 2]
```diff
@@ -581,23 +581,22 @@ func (daemon *Daemon) newContainer(name string, config *runconfig.Config, imgID
...
-		hostConfig:      &runconfig.HostConfig{},
-		ImageID:         imgID,
-		NetworkSettings: &network.Settings{},
-		Name:            name,
-		Driver:          daemon.driver.String(),
-		ExecDriver:      daemon.execDriver.Name(),
-		State:           NewState(),
-		execCommands:    newExecStore(),
-	}
+	container := &Container{}
```
Comment (tiborvass) [inline daemon/daemon.go]: same question here

File: daemon/daemon.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/daemon.go]: I agree, but there's a good reason. When I split out the Container structure into common and platform specific parts, everything is good in go except for the constructor "container:=&Container{field: value, etc}.". Accessors are fine, but the constructor will fail compilation. Hence the need to rework it to be field by field. Sucks, I know. As far as I can tell, it's a limitation in go.

File: daemon/daemon.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/daemon.go]: Same answer here :)

File: daemon/daemon.go (hunk 1 above)
Comment (tiborvass) [inline daemon/daemon.go]: Ohhh I see, what you need is this:
```
container := &Container{
    CommonContainer: CommonContainer{
        State: NewState()
    },
    root: daemon.containerRoot(id),
    execCommands: newExecStore(),
}
```

Comment (jhowardmsft) [inline daemon/daemon.go]: Ahhhh, lightbulb moment. That makes sense. I just learnt more about go. Will verify and address in the morning, Thanks!

Comment (jhowardmsft) [inline daemon/daemon.go]: That's fixed. Thanks for the pointer in the right direction :)

**Comment by jhowardmsft** [comment]
@tiborvass Fixed constructors. Thanks for the pointer to the right syntax.

**Comment by tiborvass** [comment]
@jhowardmsft awfully sorry but needs a rebase

**Comment by jhowardmsft** [comment]
Rebased

**Comment by jfrazelle** [comment]
ah so sorry needs rebase

**Comment by jfrazelle** [comment]
you can ping me after :)

**Comment by jhowardmsft** [comment]
Again??! Wow, things change fast around here :)

**Comment by jhowardmsft** [comment]
Rebased

File: daemon/container_windows.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		return nil
+	}
+
+	var (
+		env *engine.Env
+		err error
+		eng = container.daemon.eng
+	)
+
+	job := eng.Job("allocate_interface", container.ID)
```
Comment (vieux) [inline daemon/container_windows.go]: the `jobs` system was removed completely I beileve

Comment (jhowardmsft) [inline daemon/container_windows.go]: Yes, you're right. This entire implementation needs reworking with libnetwork in future. I've added a TODO instead.

**Comment by jhowardmsft** [comment]
Third time lucky on rebasing :)

File: daemon/container.go [hunk 4]
```diff
@@ -89,26 +79,21 @@ type Container struct {
...
 	RestartCount             int
 	UpdateDns                bool
 
 	// Maps container paths to volume paths.  The key in this is the path to which
 	// the volume is being mounted inside the container.  Value is the path of the
 	// volume on disk
-	Volumes map[string]string
-	// Store rw/ro in a separate structure to preserve reverse-compatibility on-disk.
-	// Easier than migrating older container configs :)
-	VolumesRW  map[string]bool
```
Comment (icecrime) [inline daemon/container.go]: Does it mean that `-v something:ro` will be unsupported on Windows?

**Comment by unclejack** [comment]
@jhowardmsft Sorry, this requires rebasing again.

File: daemon/container.go (hunk 4 above)
Comment (jhowardmsft) [inline daemon/container.go]: For the initial release, yes that is the current plan.

**Comment by jhowardmsft** [comment]
@unclejack Rebased. Waiting for Jenkins to complete.

**Comment by icecrime** [comment]
LGTM
